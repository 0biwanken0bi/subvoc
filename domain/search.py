from collections import Counter


def search(api, query, count=20):
    subtitles = api.find_by_query(query)

    movie_downloads = Counter()
    movie_by_id = {}
    for s in subtitles:
        if s.media.kind == 'movie':
            media_id = s.media.id
            movie_by_id[media_id] = s.media
            movie_downloads[media_id] += s.downloads

    popular_movies = [movie_by_id[m[0]] for m in movie_downloads.most_common(count)]
    return popular_movies
