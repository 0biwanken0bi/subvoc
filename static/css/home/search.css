.search {
    margin-top: 2.5rem;
    max-width: 60rem;
}

.search .search-wrapper {
    display: flex;
}


/* BAR */

.search .searchbar {
    background-color: #fff;
}


/* BUTTON */

.search .searchbar:focus {
    border-color: transparent;
}

.search .searchbutton {
    font-weight: normal;
    margin-left: 1rem;
}


/* BAR + BUTTON */

.search .searchbutton,
.search .searchbar {
    font-size: 2rem;

    height: 4rem;
    line-height: 4rem;
}


/* RESULT */

.search .search-result {
    margin-top: 2rem;
    font-weight: normal;
    font-size: 0.6em;
}
