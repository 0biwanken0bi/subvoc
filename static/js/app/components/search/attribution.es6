const Attribution = () => 
    <footer className="credit">
        Posters provided by <a href="https://fanart.tv">fanart.tv</a>
    </footer>;


export { Attribution }