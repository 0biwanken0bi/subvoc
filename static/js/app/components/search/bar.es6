import { debounce } from '../../util/debounce.es6';


const SearchBar = ({ onSearch }) => {
    const debouncedSearch = debounce((e) => onSearch(e.target.value), 500);

    return <div className="search">
        <div className="search-input" />
            <div className="search-wrapper">
            <input type="text"
                    className="searchbar"
                    name="q"
                    autoFocus
                    autoComplete="off"
                    onChange={(e) => {
                        e.persist();
                        debouncedSearch(e);
                    }}
                    placeholder="Search movie ..." />
        </div>
    </div>;
};


export { SearchBar };