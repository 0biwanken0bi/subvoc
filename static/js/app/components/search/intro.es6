const Intro = () =>
    <div className="step-by-step">
      <ol>
        <li>
          <span>
            <strong>search</strong> for a movie
          </span>
        </li>
        <li>
          <span>
            <strong>browse</strong> its vocabulary
          </span>
        </li>
        <li>
          <span>
            <strong>learn</strong> new words
          </span>
        </li>
      </ol>
    </div>;


export { Intro };
