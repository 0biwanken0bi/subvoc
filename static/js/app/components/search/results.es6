import $ from 'jquery';
import Slider from 'react-slick';

import { Attribution } from './attribution.es6';


const SearchResultItem = ({ item, onSelect }) =>
    <div className="search-result-item" onClick={() => onSelect(item.id)}>
        <a className="header">
            { item.poster_url
                ? <img className="poster" src={item.poster_url} />
                : <img className="poster empty" src="/static/img/placeholder.png" /> }
        </a>

        <div className="footer">
            { item.title }
        </div>
    </div>;


const SearchResults = ({ items, onSelect }) => {
    const slickSettings = {
        'infinite': false,
        'slidesToShow': 4,
        'slidesToScroll': 4,
        'responsive': [{
            'breakpoint': 1024,
            'settings': {
                'slidesToShow': 4,
                'slidesToScroll': 4,
            }
        }, {
            'breakpoint': 600,
            'settings': {
                'slidesToShow': 3,
                'slidesToScroll': 3
            }
        }, {
            'breakpoint': 480,
            'settings': {
                'slidesToShow': 2,
                'slidesToScroll': 2
            }
        }]
    };

    return <div>
        <div className="search-result-wrapper">
            <div className="search-result">
                { items.length === 0 
                    ? <div className="empty"> No movie was found. </div>
                    : <Slider {...slickSettings}>
                        { $.map(items, item => <div key={item.id}> 
                            <SearchResultItem item={item} onSelect={onSelect} /> 
                        </div> )}
                    </Slider>
                }
            </div>
        </div>
        { items.length !== 0 ? <Attribution /> : <span/> }
    </div>;
}


export { SearchResults }