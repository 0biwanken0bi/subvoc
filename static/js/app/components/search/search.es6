import $ from 'jquery';
import React from 'react';

import { Intro } from './intro.es6';
import { Nav } from '../nav.es6';
import { SearchBar } from './bar.es6';
import { SearchResults } from './results.es6';

import { Spinner } from '../util/spinner.es6';


class Search extends React.Component {

    constructor() {
        super();
        this.state = { items: undefined };
    }

    handleSearch(query) {
        this.setState((prevState) => {
            if (prevState.searchXHR) {
                prevState.searchXHR.abort();
            }

            prevState.searchXHR = this.searchMovie(query);
        });
    }

    render() {
        const { onSelect } = this.props;
        return <div>
            <Nav />
            <section className='container'>
                <h1 className="heading">
                    Use movies to discover new vocabulary.
                </h1>
                <SearchBar onSearch={(q) => this.handleSearch(q)} />
                { this.state.searchXHR 
                    ? <Spinner big={true} />
                    : this.state.items === undefined 
                        ? <Intro />
                        : <SearchResults items={this.state.items} 
                                         onSelect={onSelect} /> 
                }                
            </section>
        </div>;
    }

    searchMovie(query) {
        if ($.trim(query) === '') {
            return;
        }

        const xhr = $.getJSON({
            url: `/api/search/${query}`,
        });

        xhr.then((res) => {
            this.setState((prevState) => {
                prevState.searchXHR = undefined;
                prevState.items = res.hits;
            });
        })
        .catch((err) => {
            if (err.statusText == 'abort') {
                return;
            }
            document.location.href = "/error";
        });

        return xhr;
    }     
}


export { Search }