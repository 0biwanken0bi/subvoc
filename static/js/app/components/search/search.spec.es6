import React from 'react';
import { shallow, render } from 'enzyme';
import td from 'testdouble';

import { Search } from './search';
import { SearchResults } from './results.es6';


describe("Search component", () => {

    test("should render", () => {
        const result = shallow(
            <Search />
        );
        
        expect(result).toMatchSnapshot();
    });

    test("should render search result", (done) => {
        const API = mockAPI();
        td.when(API.searchMovie('QUERY')).thenResolve({
            data: {
                hits: ['A', 'B', 'C'],
            }
        });
        
        const wrapper = shallow(<Search api={API} />);
        wrapper.find('SearchBar').simulate('search', 'QUERY');
        
        process.nextTick(() => {
            expect(wrapper.update()).toMatchSnapshot();
            done();
        });
    });
});


function mockAPI() {
    return { searchMovie: td.function() };
}