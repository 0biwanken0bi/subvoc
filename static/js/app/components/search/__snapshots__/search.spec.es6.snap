// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render 1`] = `
<div>
  <Nav />
  <section
    className="container"
  >
    <h1
      className="heading"
    >
      Use movies to discover new vocabulary.
    </h1>
    <SearchBar
      onSearch={[Function]}
    />
    <Intro />
  </section>
</div>
`;

exports[`Search component should render search result 1`] = `
<div>
  <Nav />
  <section
    className="container"
  >
    <h1
      className="heading"
    >
      Use movies to discover new vocabulary.
    </h1>
    <SearchBar
      onSearch={[Function]}
    />
    <SearchResults
      items={
        Array [
          "A",
          "B",
          "C",
        ]
      }
    />
  </section>
</div>
`;
